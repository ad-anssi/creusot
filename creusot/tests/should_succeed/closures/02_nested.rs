fn nested_closure() {
    let a = true;
    let a = (|| {
        let omg = || a;
        (omg)()
    })();
}
