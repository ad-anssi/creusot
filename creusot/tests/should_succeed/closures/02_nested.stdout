module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
end
module C02Nested_NestedClosure_Closure0_Closure0_Interface
  type c02nested_nestedclosure_closure0_closure0   
  use prelude.Prelude
  val c02Nested_NestedClosure_Closure0_Closure0 (_1 : c02nested_nestedclosure_closure0_closure0) (_2 : ()) : bool
end
module C02Nested_NestedClosure_Closure0_Closure0
  type c02nested_nestedclosure_closure0_closure0  = 
    | C02Nested_NestedClosure_Closure0_Closure0 bool
    
  use prelude.Prelude
  let rec cfg c02Nested_NestedClosure_Closure0_Closure0 (_1 : c02nested_nestedclosure_closure0_closure0) (_2 : ()) : bool
    
   = 
  var _0 : bool;
  var _1 : c02nested_nestedclosure_closure0_closure0;
  {
    _1 <- _1;
    goto BB0
  }
  BB0 {
    assume { (fun x -> true) _0 };
    _0 <- (let C02Nested_NestedClosure_Closure0_Closure0 a = _1 in a);
    assume { (fun x -> true) _1 };
    return _0
  }
  
end
module Core_Ops_Function_FnOnce_Output
  type self   
  type args   
  type output   
end
module Core_Ops_Function_FnOnce_CallOnce_Interface
  type self   
  type args   
  clone Core_Ops_Function_FnOnce_Output as Output0 with type self = self, type args = args
  val call_once [@cfg:stackify] (self : self) (args : args) : Output0.output
    requires {false}
    
end
module Core_Ops_Function_FnOnce_CallOnce
  type self   
  type args   
  clone Core_Ops_Function_FnOnce_Output as Output0 with type self = self, type args = args
  val call_once [@cfg:stackify] (self : self) (args : args) : Output0.output
    requires {false}
    
end
module Core_Ops_Function_Fn_Call_Interface
  type self   
  type args   
  use prelude.Prelude
  clone Core_Ops_Function_FnOnce_Output as Output0 with type self = self, type args = args
  val call [@cfg:stackify] (self : self) (args : args) : Output0.output
    requires {false}
    
end
module Core_Ops_Function_Fn_Call
  type self   
  type args   
  use prelude.Prelude
  clone Core_Ops_Function_FnOnce_Output as Output0 with type self = self, type args = args
  val call [@cfg:stackify] (self : self) (args : args) : Output0.output
    requires {false}
    
end
module C02Nested_NestedClosure_Closure0_Interface
  type c02nested_nestedclosure_closure0   
  use prelude.Prelude
  val c02Nested_NestedClosure_Closure0 (_1 : c02nested_nestedclosure_closure0) (_2 : ()) : bool
end
module C02Nested_NestedClosure_Closure0
  type c02nested_nestedclosure_closure0  = 
    | C02Nested_NestedClosure_Closure0 bool
    
  use prelude.Prelude
  clone C02Nested_NestedClosure_Closure0_Closure0 as Closure00
  let rec cfg c02Nested_NestedClosure_Closure0 (_1 : c02nested_nestedclosure_closure0) (_2 : ()) : bool = 
  var _0 : bool;
  var _1 : c02nested_nestedclosure_closure0;
  var omg_2 : Closure00.c02nested_nestedclosure_closure0_closure0;
  var _3 : bool;
  var _4 : Closure00.c02nested_nestedclosure_closure0_closure0;
  var _5 : ();
  {
    _1 <- _1;
    goto BB0
  }
  BB0 {
    _3 <- (let C02Nested_NestedClosure_Closure0 a = _1 in a);
    assume { (fun x -> true) _1 };
    omg_2 <- Closure00.C02Nested_NestedClosure_Closure0_Closure0 _3;
    _4 <- omg_2;
    assume { (fun x -> true) omg_2 };
    _5 <- ();
    _0 <- Closure00.c02Nested_NestedClosure_Closure0_Closure0 _4 _5;
    goto BB1
  }
  BB1 {
    return _0
  }
  
end
module C02Nested_NestedClosure_Interface
  val nested_closure [@cfg:stackify] () : ()
end
module C02Nested_NestedClosure
  use prelude.Prelude
  clone C02Nested_NestedClosure_Closure0 as Closure00
  let rec cfg nested_closure [@cfg:stackify] () : () = 
  var _0 : ();
  var a_1 : bool;
  var a_2 : bool;
  var _3 : Closure00.c02nested_nestedclosure_closure0;
  var _4 : Closure00.c02nested_nestedclosure_closure0;
  var _5 : bool;
  var _6 : ();
  {
    goto BB0
  }
  BB0 {
    a_1 <- true;
    _5 <- a_1;
    assume { (fun x -> true) a_1 };
    _4 <- Closure00.C02Nested_NestedClosure_Closure0 _5;
    _3 <- _4;
    assume { (fun x -> true) _4 };
    _6 <- ();
    a_2 <- Closure00.c02Nested_NestedClosure_Closure0 _3 _6;
    goto BB1
  }
  BB1 {
    assume { (fun x -> true) a_2 };
    _0 <- ();
    return _0
  }
  
end
